File: backend/schemas/regions.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# /regions/stats — strongly typed contract
# ---------------------------------------------------------------------------

class RegionStatsResponse(BaseModel):
    median_deaths: int | None = None
    median_injuries: int | None = None
    median_affected: int | None = None
    median_damage_000usd: int | None = None
    n_events: int
    deaths_coverage: float
    injuries_coverage: float
    affected_coverage: float
    damage_coverage: float
    data_source: str          # "country" | "region" | "global"
    country_used: str | None = None


# ---------------------------------------------------------------------------
# Supporting sub-models for complex passthrough responses
# ---------------------------------------------------------------------------

class ContinentEntry(BaseModel):
    total_events: int
    top_disaster: str
    median_deaths: int | None = None
    median_damage_000usd: int | None = None


class SecondaryDisasterEntry(BaseModel):
    type: str
    count: int


class TimeseriesYearEntry(BaseModel):
    year: int
    events: int
    deaths: int | None = None
    affected: int | None = None
    damage_000usd: int | None = None


class TimeseriesDecadeEntry(BaseModel):
    decade: int
    events: int
    deaths: int | None = None
    affected: int | None = None
    damage_000usd: int | None = None


# ---------------------------------------------------------------------------
# Type aliases for passthrough JSON endpoints
# Routers return these dicts directly without response_model validation —
# the aliases document the contract for the frontend TypeScript types.
# ---------------------------------------------------------------------------

# {decades: list[int], "Flood": list[int], ...}  — keys are disaster types + "decades"
TrendsResponse = dict[str, Any]

# {continent: {total_events, top_disaster, median_deaths, median_damage_000usd}}
ContinentStatsResponse = dict[str, ContinentEntry]

# {disaster_type: float}  — e.g. {"Flood": 0.26, "Earthquake": 0.17}
InsuranceGapResponse = dict[str, float]

# {disaster_type: list[int]}  — month numbers 1–12
SeasonalPeaksResponse = dict[str, list[int]]

# {disaster_type: [{type: str, count: int}]}
SecondaryDisastersResponse = dict[str, list[SecondaryDisasterEntry]]


class TimeseriesResponse(BaseModel):
    by_year: dict[str, list[TimeseriesYearEntry]]
    by_decade: dict[str, list[TimeseriesDecadeEntry]]


# ---------------------------------------------------------------------------
# /regions/risk-map — heat-points for the public Leaflet map
# ---------------------------------------------------------------------------

class RiskMapPoint(BaseModel):
    """One historical disaster event with valid lat/lon and a composite risk
    score in [0, 100]. risk_score uses the CLAUDE.md formula with the
    probability term fixed at 1.0 (these are historical events, not
    predictions). Pre-sampled offline by scripts/generate_emdat_stats.py
    (3 of the 8 EM-DAT types — Drought, Wildfire, Extreme temperature — are
    omitted because EM-DAT has no point coordinates for them)."""
    lat:           float
    lon:           float
    risk_score:    float
    disaster_type: str


RiskMapResponse = list[RiskMapPoint]

File: frontend/types/regions.ts
// Mirrors backend/schemas/regions.py + observed live shapes from /regions/*.

// /regions/stats — RegionStatsResponse
export interface RegionStats {
  median_deaths:        number | null
  median_injuries:      number | null
  median_affected:      number | null
  median_damage_000usd: number | null
  n_events:             number
  deaths_coverage:      number
  injuries_coverage:    number
  affected_coverage:    number
  damage_coverage:      number
  data_source:          "country" | "region" | "global"
  country_used:         string | null
}

// /regions/countries — continent→country picker with fixed centroids.
// `name` is the exact EM-DAT country string (sent back on predictions so the
// country-tier impact lookup hits); `label` is the cleaned display string.
export interface CountryEntry {
  name:     string
  label:    string
  iso:      string
  lat:      number
  lon:      number
  n_events: number
}
export interface CountryDefault {
  continent: string
  name:      string
  label:     string
  lat:       number
  lon:       number
}
export interface CountriesResponse {
  default:    CountryDefault
  continents: Record<string, CountryEntry[]>
}

// /regions/risk-map — list of historical events with valid in-range lat/lon
// and a composite risk score in [0, 100]. Mirrors backend RiskMapPoint.
export interface RiskMapPoint {
  lat:           number
  lon:           number
  risk_score:    number
  disaster_type: string
}

// /regions/trends — {decades: number[], "<DisasterType>": number[]}
// Confirmed live: {"decades":[1950..2020], "Flood":[81,155,...], ...}
export interface TrendsData {
  decades: number[]
  // disaster-type keys: counts per decade — same length as `decades`
  [disasterType: string]: number[]
}

// /regions/continent-stats — {continent: ContinentEntry}
export interface ContinentEntry {
  total_events:         number
  top_disaster:         string
  median_deaths:        number | null
  median_damage_000usd: number | null
  events_by_type?:      Record<string, number>  // per-type event counts (added Phase 5 polish)
}
export type ContinentStats = Record<string, ContinentEntry>

// /regions/insurance-gap — {disaster_type: ratio}
export type InsuranceRatios = Record<string, number>

// /regions/seasonal-peaks — {disaster_type: month_numbers}
export type SeasonalPeaks = Record<string, number[]>

// /regions/secondary-disasters — {disaster_type: SecondaryDisasterEntry[]}
export interface SecondaryDisasterEntry {
  type:  string
  count: number
}
export type SecondaryDisasters = Record<string, SecondaryDisasterEntry[]>

// /regions/timeseries — TimeseriesResponse
export interface TimeseriesYearEntry {
  year:           number
  events:         number
  deaths:         number | null
  affected:       number | null
  damage_000usd:  number | null
}
export interface TimeseriesDecadeEntry {
  decade:         number
  events:         number
  deaths:         number | null
  affected:       number | null
  damage_000usd:  number | null
}
export interface TimeSeriesData {
  by_year:               Record<string, TimeseriesYearEntry[]>
  by_decade:             Record<string, TimeseriesDecadeEntry[]>
  by_continent_decade?:  Record<string, Record<string, TimeseriesDecadeEntry[]>>
}
